test/interpreter_functional/plugins/kbn_tp_run_pipeline/public/components/main.js

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React from 'react';
+import { unmountComponentAtNode } from 'react-dom';
+import {
+  EuiPage,
+  EuiPageBody,
+  EuiPageContent,
+  EuiPageContentHeader,
+} from '@elastic/eui';
+
+import { RequestAdapter } from 'ui/inspector/adapters/request';
+import { DataAdapter } from 'ui/inspector/adapters/data';
+import { runPipeline } from 'ui/visualize/loader/pipeline_helpers';
+
+// todo: registries must be taken out of package
+import { registries } from '@kbn/interpreter/public';
+
+class Main extends React.Component {
+
+  chartDiv = React.createRef();
+  exprDiv = React.createRef();
+
+  constructor(props) {
+    super(props);
+
+    window.runPipeline = async (expression, context = {}, initialContext = {}) => {
```

Comment:
What do you think, instead of attaching this via the global window object, doing something like creating a service to work off this app plugin.  Give the expDiv a `data-test-subj`, and also create an input box for the context (this would actually then be a pretty neat plugin by itself to play around with manually triggering the interpreter and expression by passing in context via the UI).  The service can then use that so it'd be something like:

```
expressionService.runPipeline = (expression, context = {}, initialContext = {}) => {
  testSubjects.setValue('exprDiv', expression);
  testSubjects.setValue('contextDiv', JSON.stringify(context));
  testSubjects.click('renderExpressionButton');
  await testSubjects.expectExists('renderingDone');
}

```

Then in this app, clicking the `renderExpressionButton` button would take the data out of the `contextDiv` and parse it back into a JSON object, and call `runPipeline` (and maybe also `renderPipelineResponse`...)

Haven't thought that out fully yet, but it seems like there could be a way to avoid using the globals.

Reply:
i was playing with this idea, so having actual text input in the plugin where you would enter expression and then click a button to execute it (or to render it)

however, it was more than 100 times slower than the current approach, thats why i went in this direction at the end, as i think its quite important that this also runs fast (multiple expressions per second)